chain/client/src/sync/block.rs

```diff
@@ -84,104 +102,95 @@ impl BlockSync {
-    fn check_state_needed(&self, chain: &Chain) -> Result<bool, near_chain::Error> {
-        let head = chain.head()?;
-        let header_head = chain.header_head()?;
+    fn check_state_needed(&self, head: &Tip, header_head: &Tip) -> bool {
+        if self.archive || !self.state_sync_enabled {
+            return false;
+        }
 
-        // If latest block is up to date return early.
-        // No state download is required, neither any blocks need to be fetched.
-        if head.height >= header_head.height {
-            return Ok(false);
+        log_assert!(head.height <= header_head.height);
+
+        // Only if the header head is more than one epoch ahead, then consider State Sync.
+        // block_fetch_horizon is used for testing to prevent test nodes from switching to State Sync too eagerly.
+        let prefer_state_sync = head.epoch_id != header_head.epoch_id
+            && head.next_epoch_id != header_head.epoch_id
+            && head.height.saturating_add(self.block_fetch_horizon) < header_head.height;
+        if prefer_state_sync {
+            debug!(
+                target: "sync",
+                head_epoch_id = ?head.epoch_id,
+                header_head_epoch_id = ?header_head.epoch_id,
+                head_next_epoch_id = ?head.next_epoch_id,
+                head_height = head.height,
+                header_head_height = header_head.height,
+                block_fetch_horizon = self.block_fetch_horizon,
+                "Switched from block sync to state sync");
         }
+        prefer_state_sync
+    }
 
-        // Don't run State Sync if header head is not more than one epoch ahead.
-        if head.epoch_id != header_head.epoch_id && head.next_epoch_id != header_head.epoch_id {
-            if head.height < header_head.height.saturating_sub(self.block_fetch_horizon)
-                && !self.archive
-                && self.state_sync_enabled
-            {
-                tracing::debug!(
-                    target: "sync",
-                    head_epoch_id = ?head.epoch_id,
-                    header_head_epoch_id = ?header_head.epoch_id,
-                    head_next_epoch_id = ?head.next_epoch_id,
-                    head_height = head.height,
-                    header_head_height = header_head.height,
-                    header_head_height_sub = header_head.height.saturating_sub(self.block_fetch_horizon),
-                    archive = self.archive,
-                    state_sync_enabled = self.state_sync_enabled,
-                    block_fetch_horizon = self.block_fetch_horizon,
-                    "Switched from block sync to state sync");
-                // Epochs are different and we are too far from horizon, State Sync is needed
-                return Ok(true);
+    // Finds the last block on the canonical chain that is in store (processed).
+    fn get_last_processed_block(
+        &self,
+        chain: &Chain,
+        last_block_hash: &CryptoHash,
```

Comment:
Why do I need to pass `last_block_hash` here? When I look at the signature, I don't expect to pass `last_block_hash`, I intend to receive it as the result

About implementation: why can't I just ask `chain.store().get_latest_known`?